dll/win32/imm32/utils.c

```diff
@@ -443,11 +448,25 @@ Imm32MakeIMENotify(HIMC hIMC, HWND hwnd, DWORD dwAction, DWORD_PTR dwIndex, DWOR
             if (pImeDpi)
             {
                 /* do notify */
-                pImeDpi->NotifyIME(hIMC, dwAction, dwIndex, dwValue);
+                TRACE("NotifyIME(%p, %lu, %lu, %p)\n", hIMC, dwAction, dwIndex, dwValue);
+                if (!pImeDpi->NotifyIME(hIMC, dwAction, dwIndex, dwValue))
+                    WARN("NotifyIME(%p, %lu, %lu, %p) failed\n", hIMC, dwAction, dwIndex, dwValue);
 
                 ImmUnlockImeDpi(pImeDpi); /* unlock */
             }
+            else
+            {
+                WARN("pImeDpi was NULL\n");
+            }
         }
+        else
+        {
+            WARN("dwThreadId was zero\n");
+        }
+    }
+    else
+    {
+        WARN("dwAction was zero\n");
     }
 
     if (hwnd && dwCommand)
```

Comment:
Should the notification be done even if pImeDpi is NULL, or dwThreadId or dwAction are zero?

Reply:
This function can do two notifications (`NotifyIME` and `WM_IME_NOTIFY`).
This `WM_IME_NOTIFY` notification should be done.